### ISSUE TITLE
Proposal: Dockerfile add INCLUDE

### ISSUE THREAD

**Issue by dysinger** [docker/docker#735]

**Comment by neglectedvalue** [comment]
That's will be great! :+1:

**Comment by ajvb** [comment]
This would be really nice, especially when combined with the ```docker build -f``` flag that has recently been merged into master. Allows for multiple dockerfiles with a core one so that they can stay DRY.

**Comment by mpareja** [comment]
-1 Include would only solve a very narrow use-case. You'll soon find yourselves needing just a slight bit more power.

This is why tools like chef, ansible and puppet have thriving ecosystems of "cookbooks/recipes", "playbooks/tasks" and "resources". They also allow you to tweak how those things are applied. While I don't want to manage everything using these tools, I do appreciate the re-usability and composability you get from a library of tools to help you get stuff on a machine - even if it is just once during a docker build.

Disclaimer, I haven't tried using something like chef *inside* the image build process, but I've certainly felt the pain of wanting to add the same functionality to multiple images and copying it around in Dockerfiles. I also don't think using parent images to DRY (don't repeat yourself) up your Dockerfiles is a good idea. Parent images remind me a lot of inheritance in object oriented programming - it's either precisely what you need or steer the hell clear of it because the coupling makes changing things very difficult.

**Comment by duglin** [comment]
@tiborvass @jfrazelle @crosbymichael 
I'd like to hear your take on this as the topic of an INCLUDE does pop up from time to time.

Personally, I don't see an issue with it since adding it doesn't break any design principle of Docker and if it makes people's lives better (or at least if they think it does :-)  ) then I'm ok with it. I don't really see it being much different than an "include" statement in a Makefile.

I would however suggest that we allow for it to be a bit more powerful by allowing env vars to be used - e.g.:
```
INCLUDE dockerfiles/os-specific.${os:-default}
```
But we can always do that as a secondary step.

**Comment by devTristan** [comment]
If you want to generate Dockerfiles with includes, conditionals, and turing completeness you can use [dockerscript](https://github.com/devTristan/dockerscript). Full disclosure: I made this because I wanted includes.

**Comment by jfrazelle** [comment]
Hello!
We are no longer accepting patches to the Dockerfile syntax as you can read about here: https://github.com/docker/docker/blob/master/ROADMAP.md#22-dockerfile-syntax

Mainly:
> Allowing the Builder to be implemented as a separate utility consuming the Engine's API will open the door for many possibilities, such as offering alternate syntaxes or DSL for existing languages without cluttering the Engine's codebase

Then from there, patches/features like this can be re-thought. Hope you can understand.

**Comment by jcoffland** [comment]
I know you guys are stubborn as hell on this topic, which is commendable, but why doesn't docker just run ``m4`` on ``Dockerfiles`` by default.  This would give docker many powerful features including ``include()`` with out changing the syntax.  ``docker build`` would need to be modified to accept some ``m4`` command line options such as ``-I`` but that's about it.

**Comment by caffinatedmonkey** [comment]
Its because we want an include directive which is smart and merges the images together. Someone correct me if I am wrong. Saying use m4 is like telling a car owner to start their car with a screwdriver. It works but its not elegant and adds more dependencies.



On Wed, Oct 21, 2015 at 1:26 PM -0700, "Joseph Coffland" <notifications@github.com> wrote:
I know you guys are stubborn as hell on this topic, which is commendable, but why doesn't docker just run ``m4`` on ``Dockerfiles`` by default.  This would give docker many powerful features including ``include()`` with out changing the syntax.  ``docker build`` would need to be modified to accept some ``m4`` command line options such as ``-I`` but that's about it.

---

**Comment by thaJeztah** [comment]
The reason this feature is closed now, is that we are working on moving the builder out of the daemon to the client. This allows other implementations (you'd even be able to create your own Dockerfile alternative), see https://github.com/docker/docker/blob/master/ROADMAP.md#22-dockerfile-syntax

Before that has arrived, we don't want to make changes to the Dockerfile syntax. Apart from that, we want to keep the Dockerfile syntax simple (we will never be able to have a one-size-fits-all solution), hence allowing other implementations that can be tailored more to specific needs.

**Comment by jcoffland** [comment]
@caffinatedmonkey I think you are assuming that ``m4`` is not good because it's old.  ``m4`` is more like a swiss army knife than a screwdriver.  But what does this have to do with starting cars?

@thaJeztah "we don't want to do it now" is a terrible reason for closing an issue.  This is exactly why you want to leave issues open.

**Comment by thaJeztah** [comment]
@jcoffland it was decided to close these issues until it's become more clear what the future of the builder/Dockerfiles will look like, but please "ping" once the builder change is implemented.

I'll also ask among the other maintainers if re-opening these requests before that is something to consider

**Comment by jcoffland** [comment]
I recommend using GitHub milestones.  You could reopen these issues and put them in a future milestone.  In my experience, this keeps issue reporting users happy because they don't see you dismissing ideas they consider important.  Yet you can still easily sort out issues which are not currently important to you.  I often create a non-versioned milestone with a name like ``unscheduled`` for this purpose.

**Comment by caffinatedmonkey** [comment]
@jcoffland I'm not assuming `m4` is not good because it's old. It is an example of design has stood the test of time. I was trying to use the car screwdriver thing as a metaphor.

**Comment by tomalakgeretkal** [comment]
I would have used this so that my 32-bit and 64-bit CentOS 6 images wouldn't be 184 identical lines with a couple added to the top of just one.

**Comment by jcoffland** [comment]
@tomalakgeretkal I use ``m4`` for exactly this purpose.  You can do so now by preprocessing your ``Dockerfile`` with something like this:

```
m4 -I . base.m4 > Dockerfile
```

It would be even better if docker either had an option to automatically preprocess with ``m4`` or better yet always did so.  Docker would need to add ``-I``, for include, to the ``docker build`` command but that's about it.  Supper easy to implement and a general solution which solves a lot of problems.

**Comment by rncry** [comment]
+1 for this (whenever it gets reopened)

**Comment by assalw** [comment]
+1 This would really be great! I need this to structure my Docker files in a sensible way.

**Comment by kujiy** [comment]
+1
WE REALLY NEED!

**Comment by avirshup** [comment]
[Docker Make](http://www.github.com/avirshup/DockerMake) might be a solution to many of the problems here. It supports both:
1. support dockerfile "inheritance", and
1. ADDs/COPYs from multiple locations.

Here's @binocarlos 's example, with files added from multiple locations. You specify the build with a YAML file:
```yaml
imagemagick:
 build_directory: /path/to/data
 build: |
   RUN apt-get install imagemagick
   ADD x.png #adds /path/to/data/x.png

redis-session:
 build_directory: /another/path/
 build: |
   RUN apt-get install redis-server
   ADD y.dat #adds /another/path/y.dat

combined:
 FROM: appserver
 requires:
  - redis-session
  - imagemagick
```

And create the final image just by running
```
docker-make.py combined
```

Full disclosure: I wrote DockerMake because I was sick of dealing with "FROM" dependencies.

**Comment by kujiy** [comment]
@avirshup Is Docker-Make a tool like [docker-compose](https://docs.docker.com/compose/) for making a docker image?

**Comment by jmcejuela** [comment]
+1

**Comment by jakirkham** [comment]
Yeah, the idea of parsing `Dockerfile`s through `m4` is abhorrent to me for exactly the [reason]( https://github.com/docker/docker/issues/735#issuecomment-150047453 ) mentioned by @caffinatedmonkey. Namely we are shipping off a feature that really belongs in the core spec. This means it will be reimplemented in various different ways and I may not be able to reuse the same strategy that someone else has as there is no official standard (worse code duplication than just setting and enforcing a standard). Not to mention, we have now added a dependency that will be hard to get on some platforms. As far as a path, we are already looking at paths for context anyways and we already have mechanisms for changing them. In short, I prefer the original proposal.

**Comment by Vanuan** [comment]
Another solution is Rocker: https://github.com/grammarly/rocker
IMO, it solves all the weak points of Dockerfile, making it suitable for development.

**Comment by rgpublic** [comment]
Rocker doesn't seem to have INCLUDE. At least I can't seem to get it to work. Surprisingly complex issue, indeed! I'm a novice Docker user and was really suprised that sth. so simple and straight-forward like an INCLUDE statement isn't - no pun intended - included. If Rocker doesn't have this as well I think I will need to write some kind of custom script to merge multiple files together, Oh, well...

**Comment by Vanuan** [comment]
@rgpublic it's called IMPORT

**Comment by rgpublic** [comment]
@Vanuan: Thank you so much for your response, but I think IMPORT does sth. different:

INFO[0000] IMPORT "includes/init.rocker"     
FATA[0000] You have to EXPORT something first in order to IMPORT 

I don't want to export sth. first or move stuff between images, I originally only wanted to split a long and confusing .docker file into multiple ones to make it easier to read and more modular. Just like the INCLUDE statement in PHP, CSS, C++ and many other languages. Isn't that what INCLUDE is all about? Strangely, they even added INCLUDE to the syntax highlighting even if it doesnt work:

https://github.com/grammarly/rocker/issues/29

This all simply turns into too much of a mess for me for such a "minor" issue. I think I'll simply write a small preprocessor script to sort this out.

**Comment by Vanuan** [comment]
Yeah, it's a different one. Sorry

**Comment by phpdude** [comment]
+1 to INCLUDE

**Comment by adw0rd** [comment]
:+1:

**Comment by boneyao** [comment]
+1

**Comment by caffinatedmonkey** [comment]
😭 Does no one realize the `+1` spam isn't necessary anymore?

**Comment by igorpejic** [comment]
+1

**Comment by fengli79** [comment]
INCLUDE +1

**Comment by bitsofinfo** [comment]
+1

**Comment by nfedyk** [comment]
+1

**Comment by peloncano** [comment]
+1

**Comment by TalBeno** [comment]
It is mind boggling that there is no way to reuse building blocks i.e. collection(s) of layers per functional intent, at arbitrary positions within a dockerfile. With INCLUDE or any equivalent solution. What am I missing?

**Comment by Vanuan** [comment]
The thing I was missing previously is that each line in Dockerfile creates a new layer. Thus you should never have more than 5-10 lines in Dockerfile to keep image minimal.

**Comment by suntong** [comment]
> Thus you should never have more than 5-10 lines in Dockerfile to keep image minimal.

Not necessary. I put in whatever necessary, then [remove the intermediate containers/layers to save space & speed things up with the `-rm` switch](https://sfxpt.wordpress.com/2013/11/17/debianubuntu-package-developing-with-docker-continued/)

**Comment by thaJeztah** [comment]
@suntong the `--rm` flag only removes the intermediate container, but does nothing to optimize your image, also there's no need to specify it, as it's actually the default:

```
 --rm                      Remove intermediate containers after a successful build (default true)
```

Intermediate _layers_ will still be added, so if (for example) you `ADD` a 100 megabyte file, and in the next step `RUN rm big-file`, your final image will still be 100 megabyte in size.

However, that discussion is separate from the discussion at hand here; an `INCLUDE` statement would only be for _convenience_ and not make any difference other than that.

**Comment by bmudda** [comment]
+1

**Comment by Vanuan** [comment]
Yeah, it would be much more convenient to create thousands of layers and monolithic apps. Personally, I wouldn't want to see such Dockerfiles in the docker ecosystem.

Let's look a the use case.

It sounds like people want multiple inheritance here, to build a **monolithic** image which uses **multiple languages and technologies**.

AFAIK, the philosophy of docker ecosystem is to **break down monolithic apps** to distinct **microservices** by technology. So that you wouldn't need to import ruby Dockerfile into nodejs Dockerfile.

That what I believe is the rationale behind the decision not to include `INCLUDE`: to enforce microservices architecture to keep Docker ecosystem a lightweight look and feel.

**Comment by suntong** [comment]
On Wed, Oct 12, 2016 at 11:27 AM, John Yani wrote:

> It sounds like people want multiple inheritance here, to build a
> *monolithic* image which uses *multiple languages and technologies*.
>
>
I don't agree. But anyway...

after all, "restricting how people use software" is what MS does best,
and what I dislike.

I believe, in essence, docker is meant to give people freedom,
not restriction.

**Comment by leethomas** [comment]
Since it doesn't look like this will ever be officially supported, I wrote a command line tool called [harbor](https://github.com/leethomas/harbor) that suffices for my personal use cases. It supports recursive INCLUDEs. Posting it here in case anyone else might find it useful.

Obviously some directives should not be included multiple times in a single Dockerfile. Like `FROM` or `CMD` for example, so take caution there.

**Comment by artheus** [comment]
+1

**Comment by josephtyler** [comment]
+1000

**Comment by vito-c** [comment]
@jfinkhaeuser 

Your suggestion is flawed in several ways:
1. It doesn't take advantage of image caching.
2. It adds two `FROM` statements and two `CMD` statements to the output
3. `/*` is used in the *.docker files as a path expressions (ie `rm -rf /var/lib/apt/lists/*`) but is interpreted as a comment and has to be escaped `/\*` (most dockerfiles don't do this).
4. A comment in a Dockerfile starts with `#` but is a preprocessing directive so you have to strip all of your Dockerfiles of comments.

I do see your point about a preprocessor though and I also think that `INCLUDE` is a bad idea.

Original Suggestion:
Makefile:
```makefile
Dockerfile: Dockerfile.in *.docker
    cpp -o Dockerfile Dockerfile.in

build: Dockerfile
    docker build -rm -t my/image .
```

Dockerfile.in:
```Dockerfile
FROM ubuntu:latest
MAINTAINER me

#include "imagemagick.docker"
#include "redis.docker"
```
imagemagick.docker: https://hub.docker.com/r/acleancoder/imagemagick-full/~/dockerfile/
redis.docker: https://github.com/docker-library/redis/blob/6cb8a8015f126e2a7251c5d011b86b657e9febd6/3.0/Dockerfile

**Comment by jfinkhaeuser** [comment]
@vito-c 

1. I don't see your first point. The generated `Dockerfile` would be the same if none of its input files change, which means image caching works just as well with as without generating the file. Maybe I misunderstand something?
1. It only adds `FROM` and `CMD` multiple times if you intend to put them into the included files. I don't think that's good form for something that's essentially a fragment of a `Dockerfile`. An `INCLUDE` statement would not change that unless it explicitly strips these from the included file.
1. So quote paths. That's good form anyway because of variable expansion and spaces: `RUN rm -rf $SOME_PATH/lists/*` might not do what you want if `$SOME_PATH` contains `/ /tmp`. Quotes are your friends with anything related to shell scripts and variables.
1. So use `/* */` for comments.

Really, these arguments make sense if you expect to just `#include` any downloaded third-party `Dockerfile`. But if you're building your own setup, you just have to adapt to the choices you've made. If your choice is to use the C preprocessor, then these are the consequences.

**Comment by thangbn** [comment]
+1

**Comment by xanview** [comment]
Is this available already?

**Comment by duglin** [comment]
@xanview nope

**Comment by Vanuan** [comment]
@xanview it is currently recommended to use FROM approach + multi stage builds:
https://github.com/moby/moby/issues/2745#issuecomment-319839656

**Comment by huajunzeng** [comment]
Multiple stage build is only a workaround and it's not the best workaround in my opinion. It brings additional burden of maintaining intermediate images and their versions and you can easily forget to rebuild those images when the codes are updated. 

The best workaround till now is still cp someLib . && docker build . && rm someLib. Or similarly you can use six8's dockerfactory. Unfortunately these don't work when you want to do docker-compose build.

**Comment by Vanuan** [comment]
> you can easily forget to rebuild those images when the codes are updated.

similarly, you can easily forget to cp someLib

**Comment by xanview** [comment]
@huajunzeng What do you mean by cp someLib && docker build . && rm someLib? - why would you do that? -- couldn't the docker file wget it or something, how does this address splitting a Dockerfile into multiple smaller chunks?

**Comment by huajunzeng** [comment]
@xanview My understanding of the whole point of this discussion is all about multiple build context. Whether to use INCLUDE, or to COPY from parent directory, etc. By using cp we can just manually pack different directory into a single build context and send to docker.

**Comment by Vanuan** [comment]
> is all about multiple build context

No, I think it's about reusing multiple Dockerfiles. E.g. when you want to have nginx and node in the same image.

> By using cp we can just manually pack different directory into a single build context and send to docker.

That's the thing, you can't copy something from an image. Consider you have nginx image:

```
FROM nginx

ADD nginx.conf /etc/nginx/nginx.conf
```

If you want to add node container in addition to nginx:

```
FROM node

ADD . /src/
WORK_DIR /src/
RUN yarn install
```

Copy command doesn't help you here anyhow. Only if you use multistage builds you can copy files between images:

```
FROM nginx

ADD nginx.conf /etc/nginx/nginx.conf
FROM node

ADD . /src/
WORK_DIR /src/
RUN yarn install

COPY --from=0 /path/to/nginx/resources /path/to/nginx/resources
```

I agree that this doesn't cover other use cases like exposed ports, entrypoints (how would you combine them?), users (tricky), etc. And you also have to know which folders contain needed files and libraries. But it's a start.

**Comment by zh794390558** [comment]
Did this feature can be accesed now?

**Comment by davesque** [comment]
Here's a use-case for `INCLUDE`:

Heroku's support for docker containers allows one to push multiple images for different process types in this manner:
https://devcenter.heroku.com/articles/container-registry-and-runtime#pushing-multiple-images

Say you have a Django app which defines some asynchronous Celery tasks.  Then you need a web process type and a worker process type.  But they should both have essentially the same runtime environment.  It would be convenient to just have your base `Dockerfile.web` which looks something like this:

```
FROM python:3.6

...

CMD ["web-entrypoint.sh"]
```

 and then have `Dockerfile.worker` look like this:

```
INCLUDE Dockerfile.web

CMD ["worker-entrypoint.sh"]
```

As it stands, Heroku provides no way to specify a different command using the same image for a different process type.  And Docker provides no way to do simple Dockerfile templating.  So I'm stuck with duplicating the same Dockerfile and changing only one line.

It's a bit surprising to see that this proposed feature has been debated for over four years now.  This is really not a complex feature.  The C and C++ programming languages have successfully used templating for decades.  I can't see how this would do much harm to the Docker philosophy.

**Comment by duglin** [comment]
@davesque you may want to continue to chat over here: https://github.com/moby/moby/pull/12749

**Comment by Stannieman** [comment]
I also felt the need for this for some multi stage dockerfiles where some of the stages are exactly the same for several dockerfiles. I know there are other ways to solve this, for example building the stages as separate images, but that's all a bit overkill for a relatively "simple" environment.

I ended up writing this PowerShell script:
`function WritePreproccessedFile {
    Param([string]$filePath, [string]$outputFilePath)

    $lines = GetLinesFromFile -filePath $filePath

    if ((Test-Path($outputFilePath)) -eq $true) {
        Remove-Item $outputFilePath
    }
    
    if ($lines.Length -gt 2 -and $lines[$lines.Length - 1] -eq '' -and $lines[$lines.Length - 2] -eq '') {
        $lines.RemoveAt($lines.Length - 1)
    }
    [System.IO.File]::WriteAllLines($outputFilePath, $lines)
}

function GetLinesFromFile {
    Param([string]$filePath)

    $lines = New-Object System.Collections.Generic.List[string]

    foreach ($line in [System.IO.File]::ReadLines($filePath)) {
        $includeFilePath = GetIncludeFilePathFromLine -parentFilePath $filePath -line $line

        if ($includeFilePath -eq $null) {
            $lines.Add($line)
            continue
        }

        $linesFromIncludeFile = GetLinesFromFile -filePath $includeFilePath

        $lines.AddRange([System.Collections.Generic.List[string]]$linesFromIncludeFile)
    }

    return $lines
}

function GetIncludeFilePathFromLine {
    Param([string]$parentFilePath, [string]$line)

    $includeKeyWord = 'INCLUDE '

    $line = $line.Trim().Trim('/').Trim('\')

    if ($line.StartsWith($includeKeyWord) -eq $false) {
        return $null
    }

    $includeFile = $line.Substring($includeKeyWord.Length)

    $includeFile = $includeFile.Trim('"', "'")
    
    if ($includeFile.Length -lt 1) {
        return $null
    }

    return GetAbsoluteFilePath -parentFilePath $parentFilePath -includedFilePath $includeFile
}

function GetAbsoluteFilePath {
    Param([string]$parentFilePath, [string]$includedFilePath)

    if ([System.IO.Path]::IsPathRooted($includedFilePath) -eq $true) {
        return $includedFilePath
    }

    $parentFileDirectory = Split-Path -Path $parentFilePath

    return "$parentFileDirectory/$includedFilePath"
}`

Call WritePreproccessedFile with the dockerfile to process and the output file as parameters.
It's fairly simple and maybe not super robust, but works perfect for me.

To include a file to your dockerfile do this:
INCLUDE path/to/../../other/file with spaces

This is recursive so you can have other includes in the included file.

**Comment by himslm01** [comment]
My use case is for reusing a block of ARGs in multiple places.

I have one Dockerfile which downloads the source and libraries for a project. That Dockerfile needs to know the versions of the libraries to download.

I have a second Dockerfile for building the source into an executable. I build from the second Dockerfile multiple times to build for various architectures, having downloaded the source only once. The second Dockerfile needs all of the ARGs of the first Dockerfile for the versions of libraries, plus extra args for build parameters for the different architectures.

Could I use a multi-stage build for this, ensuring that the source is only downloaded once per run? 

Other requirements: I run this process from a CI tool once a day, I always want to start by invalidating the cashe once - so that all of the source is downloaded afresh, but only on the first run for the first architecture.

**Comment by rodvdka** [comment]
+1

**Comment by bg172** [comment]
+1

**Comment by kkm000** [comment]
I am mildly missing this feature for testing multiple platfrorms. This would minimize maintenance, I absolutely do not want to copy/paste common stuff. Currently, bash and cat is my preprocessor. Not super important, but would be nice to have. Nothing Turing-complete here:
```
FROM opensuse/amd64:13.2
??INCLUDE?? prereq-suse.inc
??INCLUDE?? common.inc
```
This pattern works for opensuse, debian, ubuntu, centos and partly Fedora ("old" Fedora/"new" Fedora single split). 

I absolutely do not buy the argument that adding the INCLUDE statement to your declarative language inevitably leads to Turing-completeness, obesity, fleas on your dog and crime in your neighborhood. This is a stretch far and thin.

**Comment by borgogelli** [comment]
+1

**Comment by anton-matosov** [comment]
Tensorflow team have implemented this functionality using an `assembler.py` python script
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/dockerfiles which assembles partial dockerfiles and the rules in spec.yml

**Comment by hholst80** [comment]
It might not be obvious to everyone how to use the multi stage builds to work around the missing INCLUDE. The example below should be fairly self explanatory. Another obvious thing is the inefficiency that the staging build steps are always done even for "production" images. 

See also https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact

```
ARG env=production

FROM nginx:alpine AS production
VOLUME /etc/letsencrypt
HEALTHCHECK CMD test -f /etc/letsencrypt/.reload && rm -f /etc/letsencrypt/.reload && kill 1 || exit 0
COPY nginx.conf /etc/nginx/nginx.conf

FROM production AS staging
COPY nginx.no-ssl.conf /etc/nginx/nginx.conf

FROM $env
```

**Comment by thaJeztah** [comment]
> Another obvious thing is the inefficiency that the staging build steps are always done even for "production" images.

Try setting `DOCKER_BUILDKIT=1`, because that's only the case when using the classic builder; when using buildkit, those stages will be skipped if they're not needed to produce the final stage

**Comment by zh794390558** [comment]
+1

**Comment by MBeijer** [comment]
+1

**Comment by vrothberg** [comment]
[Buildah](https://github.com/containers/buildah) supports including other Dockerfiles since 2018. To avoid extending the Dockerfile syntax and remain compaitble, it's using the C pre-processor to assemble the Dockerfile (see the [Buildah docs](https://github.com/containers/buildah/blob/master/docs/buildah-bud.md)). Buildah only runs `cpp` on a Dockerfile if it's suffix is `.in`. Other Dockerfiles can be included via the `#include $PATH` directive.

Certainly, the C pre-processor allows writing macros and we can do horrible things with it, but it's used an understood since the 60s and present on pretty much all systems, giving a certain degree of portability. Note that this is method does not require Buildah at all and it can be used by pretty much everybody trying to implement include-semantics for Dockerfiles. Simply run `cpp -E -iquote . Dockerfile.in` and the output can be parsed by any Dockerfile-compatible tool.

**Comment by tanzislam** [comment]
If you're using the preprocessor in the latest versions of GCC for this purpose, then you'd need the following switches:
 - `-traditional` to prevent GCC from removing trailing backslashes in non-directive lines.
 - `-undef` to prevent replacing any occurrences of the word "linux" in your Dockerfile with "1", etc.

**Comment by ringerc** [comment]
Spell it as `FROM PATH` if you're unhappy with calling it "include". But please, please consider adding this, the lack is painful and forces higher levels of a stack to do direct Dockerfile munging when they should be able to consider a single Dockerfile + context dir to be an authoratative source.

**Comment by dougmoscrop** [comment]
I still think this could be accomplished in a way that is (hopefully) compatible with the current APIs as long as the containers shared an ancestry, i.e. someone should be able to do

FROM alpine
INCLUDE alpine/xyz
INCLUDE alpine/abc
blah blah

at that point the alpine/xyz just becomes a way of factoring out commons steps.  could xyz install something and abc needs a different version? of course, but that's no different than building a container that has two separate package install steps - in fact, it becomes that, it's just about DRYing the canonical build steps for e.g. java, node, etc.

**Comment by perllaghu** [comment]
Another use case:

You are building multiple variations on a theme:

```
    FROM <OS-of-choice>
    RUN <update-OS>
    INCLUDE <install common libraries set>
    RUN <install variation libraries>
    INCLUDE <install common config set>
```

For **me** - when I update the common libraries, I have to edit 11 files, not just one

For **me** - when I update the common config, I have to edit 11 files, not just one.

**Multi Stage builds is not always the answer**

Multi-stage builds are great, and they can reduce the foot-print of the final image - eg:

```
    FROM node:12
    ADD <files>
    WORKDIR /path
    RUN yarn install
    RUN yarb build

    FROM nginx
    COPY <config>
    COPY --from=0 /<path>/dist /usr/share/nginx/html
```
This throws away _all_ the rubbish from the yarn build, and only takes the resultant built tree.

.... but this only works when the thing from the first stage is whole & complete & isolatable - code libraries often aren't: Some compile against OS libraries; some build configuration details based on paths.... 

Being able to say "insert these lines into the Dockerfile here" covers this different use-case.

**Comment by klakegg** [comment]
This is a crucial feature when moving forward with multiple stages/targets/architectures, and has finally grown into an itch I simply had to scratch.

My humble solution is preprocessing of the Dockerfile so a few lines of code may do some imports and update some references in the imported instructions. It is available here:
https://github.com/klakegg/dockerfile-import

This is not a perfect solution, however I guess it may relieve some of the pain until it is fixed in Docker.

**Comment by tantal92** [comment]
another scenario is when you have many images, and in every image you need to set up the same set of ENV variables;
(e.g.: INCLUDE file_with_my_envs)

**Comment by edrevo** [comment]
I have implemented a custom buildlkit frontend that supports a new "INCLUDE" scenario. You can use it like this on recent Docker builds:

```Dockerfile
# syntax = edrevo/dockerfile-plus

FROM alpine

INCLUDE+ Dockerfile.common

ENTRYPOINT [ "mybin" ]
```

More info at https://github.com/edrevo/dockerfile-plus

**Comment by cseufert** [comment]
Even if it was just to be able to reference common build files for multi-stage builds

eg
```Dockerfile
FROM build-env --dockerfile=./Dockerfile.build-env as build
```

**Comment by eng-dibo** [comment]
a use case is when you need to create an image from a base image, in this case you need to build the base image first.

but if you want to use a continous deployment service that needs to build your image when pushing a new commit to the head branch of your code base,  like 'Google cloud run', it requires you to provide it with the Dockerfile path in your code base and it do the build process.

you don't have the option to build the base image first, before building the app image.

here you need to `include` the base Dockerfile rathar than building from the base image.

**Comment by hexcowboy** [comment]
Here's a new project I'm working on with a great use case for this:
- https://github.com/hexcowboy/jackbox/tree/main/scripts

**Comment by endrift** [comment]
It's really telling the number of different use cases that have been presented here and genuine grievances, which have been brushed off with non-obvious workarounds, forks, and other "are we out of touch? No, it's the user who is wrong" nonsense endemic to the tech industry these days. The people downvoting all the +1s are just being passive aggressive without considering that maybe the reasons are valid. It's really disappointing, even if not surprising, just how little people involved with the project seem to care.

I have a use case. So what did I do? Copy pasted a bunch of stuff and now I have to update 4 different files any time the stuff in the middle changes. Thanks, Docker.

**Comment by IamTheCarl** [comment]
Many of the proposed workarounds would work for me, except I'm maintaining both x86 and ARM images which have minor differences in their initial setup but then are otherwise identical. Different userspace drivers implemented as runtime dependencies need to be installed on both images.

My current solution is to have a common directory with all of the needed assets and a setup script. I copy that whole directory into the images during the build, and then remove the directory once the build is complete. This works but I lose all of that sweet juicy caching Docker would normally provide.

I can't just copy from one image into the other like the developers and documentation suggest because I'd be copying x86 binaries into an ARM image, or vise versa.
It also plays poorly with apt-get since packages won't be properly "installed". It results in packages being unremovable or unable to upgrade. It's not a big deal since it's rare such a thing would be useful in a Docker container, but it makes this workaround feel like a real nasty hack.

The fear of Docker files becoming "turning complete" is irrelevant because, with their de facto standard of using shell commands, they're already turning complete enough. You can write conditional statements and loops. Only the copy directive can't be controlled with these conditionals, and that's easy enough to workaround. Just have a conditional delete after or use an argument to change the path it copies from.

A way to copy assets from one image to another is not what we're asking for here.
We are asking for a way to copy build steps into a build script at build time.

**Comment by alexpirine** [comment]
I see users fall back to using "FROM" since there is no way to "INCLUDE" local files.

But since the "FROM" instruction falls back to downloading from the public repository, that's quite a security concern.

I don't really understand why we need to theorize the concern that Dockerfile syntax can become turing-complete if we enrich it we an INCLUDE instruction, while we have very obvious practical, and real, shortcomings.

And not everybody has `cpp` already installed, nor needs to understand how its preprocessor is working.

**Comment by wildone** [comment]
> Can someone give me a few examples on how they would use this? ping @ptone @keeb @dysinger

I have a 1000 likes docker file and with lots of FROM base as build-xxx layers, having each of those layers in an individual file would make the changes more manageable.

**Comment by caniko** [comment]
The controvosy of this issue is moot. `INCLUDE` is just syntactic sugar that makes sense, please implement.

**Comment by kevrat** [comment]
+1
Monorepo with a lot of services with the same Dockerfiles could be more convenient with this feature.
So far we have the [workaround](https://github.com/edrevo/dockerfile-plus), but it would be nice to have it in docker by default

**Comment by donn** [comment]
It's now becoming even more of a problem given that the `docker-container` driver of buildx cannot access the local image cache, meaning a temporary common image has to be pushed to Docker Hub to be used in a later build. I urge the maintainers to reconsider.

**Comment by mccolljr** [comment]
It has been almost 10 years since this issue was originally opened.
It has seen overwhelming support from users in the issue thread. 
It seems that the only "real" counterpoint is that... this might make the Dockerfile description language turing complete?
It also appears that the idea of `FROM ./relative/path` has been rejected?

Am I understanding the state of affairs correctly? If so, this is both surprising and disappointing.

**Comment by endrift** [comment]
It's quite clear that upstream doesn't care at all, despite the outpouring of public requests. The only thing at this point that surprises me is that they haven't made their disdain clear by locking the issue.

**Comment by perllaghu** [comment]
> It's quite clear that upstream doesn't care at all, despite the outpouring of public requests. The only thing at this point that surprises me is that they haven't made their disdain clear by locking the issue.

.... or, perhaps, they're waiting for someone _else_ to do some coding & create a `Pull Request` with [some part of] this solved?

**Comment by endrift** [comment]
I don't think anyone's gonna give that much free labor to an open source project that's commercially backed already.

**Comment by IamTheCarl** [comment]
> Not to sound negative here (pun intended), but -1.
> 
> I completely get the use cases for an include statement. But then I also get the need for parametrized Dockerfiles, and then for conditionals, etc. Continue down that path, and you'll end up implementing a programming language in Dockerfiles, which may even become turing complete. The cynicism in that statement is free of charge, by the way ;)
> 
> Why not use a preprocessing step instead? You don't even have to program your own, you could use m4 (used in autotools for this purpose) or the C preprocessor (used in IMake, which does a similar job as autotools but is pretty much defunct these days).
> 
> Makefile:
> 
> ```makefile
> Dockerfile: Dockerfile.in *.docker
>   cpp -o Dockerfile Dockerfile.in
> 
> build: Dockerfile
>   docker build -rm -t my/image .
> ```
> 
> Dockerfile:
> 
> ```
> FROM ubuntu:latest
> MAINTAINER me
> 
> #include "imagemagick.docker"
> #include "redis.docker"
> ```
> 
> Run `make` and it'll re-build the Dockerfile if any input file changed. Run `make build` and it'll re-build the Dockerfile if any input file changed, and continue to build the image.
> 
> Look ma, no code!

I got desperate and decided to give this a try, and then quickly realized that cpp can't tell the difference between a C preprocessor directive and a Docker comment.

Has anyone actually tried this method and gotten it to work?

**Comment by tanzislam** [comment]
@IamTheCarl You need to use C-style comments (`/* ... */`) instead.

**Comment by IamTheCarl** [comment]
That's just bonkers... I didn't know Docker supports C-style comments or does cpp just strip those out before they can be a problem?

Well, I tried out [edrevo's tool](https://github.com/edrevo/dockerfile-plus) which not only was easier to set up than expected (you don't need to install anything on the host) and it gave me all the benefits I was hoping to find.

**Comment by endrift** [comment]
I believe CPP strips them out.

**Comment by vovech** [comment]
Please don't use edrevo's. Not mantained and doesn't work with buildkit (which will became a default soon):

Paralel building completelly messed up:
https://github.com/edrevo/dockerfile-plus/issues/13

Currently I didn't find any acceptable solution.

**Comment by sffc** [comment]
If you are using Podman, according to [the podman-build man page](https://manpages.ubuntu.com/manpages/impish/man1/docker-build.1.html), you can build from a `Containerfile.in` that gets preprocessed into a `Containerfile`, presumably with inspiration from how a `Makefile.in` gets preprocessed into a `Makefile`.

**Comment by c33s** [comment]
working example:

`Containerfile.in`:
```
FROM alpine:latest

#include "Containerfile.defaults"

RUN ls -lsa
```

`Containerfile.defaults`:
```
RUN echo "hello world from included file"
```

```
podman build -f Containerfile.in .
```

**Comment by amohar** [comment]
I think the reasons for not implementing this feature are completely bonkers. I understand that having functionality like INCLUDE can lead to bad coding practices, as @jfinkhaeuser mentioned (like Dockerfile not being self-contained), but then we should simply throw out computing altogether, as there is not a single programming language or scripting language that doesn't introduce the possibility of bad practices. Not having the functionality implemented is not the way to mitigate them.

I, for instance, am having an issue where I want to have many (a few dozen) different Dockerfiles for many different containers, but each starts with the same base installation procedure. Then I need the possibility to do additional env setups for each one. And I need this to work locally and in the cloud. You can't convince me that having such a setup separately for each container is a good practice because it is not; the potential for human error while updating the base part is huge.

So now I'm left with having to implement one of the hacks. I either have to have a bunch of repositories that cost money for a bunch of base images that I'm not actually using or figure out a hack that will create and drop them during the build process. Or I have to figure out a 3rd-party way to preprocess them, which is a pain in the ass doing for both local and cloud. And in the end, I'm ending up with a hack instead of an accepted practice for doing an obviously often-seen scenario.

My opinion is that the reason for not implementing this is just the basic team's unwillingness to do it instead of some logical reasoning why implementing this would be a bad idea because there would be some official explanation. And if this is not the case, as somebody said this is both surprising and disappointing.

**Comment by jfinkhaeuser** [comment]
It's kind of amusing that I'm being quoted here in 2023. While I believe my argument is sound enough, also today. But also, it's not as if I care enough. 

Trust me on one thing, though, nobody listens to me. I doubt my argument is the reason this is still not done a decade later.

**Comment by creative-resort** [comment]
It looks, like an equivalent of this functionality is meanwhile provided by Docker "Bake" and "BuildKit", outside of Dockerfiles.
https://www.docker.com/blog/dockerfiles-now-support-multiple-build-contexts/
(Create Build Pipelines by Linking bake Targets)

Reference:
https://docs.docker.com/build/bake/build-contexts/#using-a-result-of-one-target-as-a-base-image-in-another-target

**Comment by gkpln3** [comment]
I've encountered the same thing with so many projects I ended up creating my own Dockerfile generation framework to provide a solution for this problem. 

Its called Sand and can be found here: https://github.com/gkpln3/Sand

**Comment by rod-dot-codes** [comment]
> I've encountered the same thing with so many projects I ended up creating my own Dockerfile generation framework to provide a solution for this problem.
> 
> Its called Sand and can be found here: https://github.com/gkpln3/Sand
> 
> You can try it
> 
> ```
> pip3 install docker-sand
> ```
> 
> Star the repo if it has helped 😀

Thanks for the tool, but I hope ten years of subscribing to this issue will not end up having been in my top 10 biggest disappointments in life - at the moment, it's probably around 30th, but the fact I'm forgetting some of my other disappointments is pushing this further to the top.

I really don't think it's adequate to work around this, especially if Podman implements the ability to include files. I mean, CI pipelines are complicated enough - why not allow us to source different parts of a multi-stage Pod Dockerfile from separate files - even in the same folder?

By chance, this issue has also served me well as a periodic Github health-check for my non-work Github email subscriptions - the flurry every 3 months or so has provided ample shaking of head time. Third place medal feels like the right emoji here.🥉

**Comment by black-snow** [comment]
Didn't go through the whole thing but let me quickly add my current use case:

I have one Dockerfile to build a DB image with a reasonable health check frequency for production.  
For my testcontainers I want a more aggressive probing, though, to wait for the health check to be green and run the tests as quickly as possible.

Now I don't want to copypaste the rest of the build file and maintain two of them - I just want to have a different healthcheck in the latter one. I also don't want to rely on any existing image (multi-stage).

**Comment by devthejo** [comment]
Hello folk's to address this need and more generally the needs of factorization on dockerfile I developed this buildkit custom syntax frontend plugin: [devthefuture/dockerfile-x](https://github.com/devthefuture-org/dockerfile-x)

just one line at the top of file `# syntax = devthefuture/dockerfile-x` and that's it !

 ```
# syntax = devthefuture/dockerfile-x

FROM ./base/dockerfile

COPY --from=./build/dockerfile#build-stage /app /app

INCLUDE ./other/dockerfile
```

**Comment by rishiraj88** [comment]
Wow! @devthejo

**Comment by devthejo** [comment]
> @devthejo the link shows 404 for me - perhaps the repo is private?

@vlad-ivanov-name nope, but the orga was, good catch, it's fixed now, thanks

**Comment by devthejo** [comment]
> Thanks for posting your work!
> 
> To me a major problem with those alternative frontends is that as far as I understand, there is no way to "layer" them on top of the official `docker/dockerfile`. So all of the additional features of the official frontend would be unavailable.

Here is how I developed this tool, the nodejs part is responsible to compile the custom syntax dockerfile to a standard dockerfile, in a superset way, then the buildkit frontend service translate this final standard dockerfile to llb (with minimal code and relying on official buildkit packages).
So if new features are developed on `docker/dockerfile` (it doesn't happen so often IMHO), there is two ways to address this, upgrade the package in the golang part of `dockerfile-x` (as I was saying there is no so much code, the maintenance cost seem very light for me).
The second way is to not use the custom syntax as frontend but precompile the Dockerfile-X to standard Dockerfile using only the nodejs part of project that is a ready to use CLI (I will add it to the documentation), you can use it via `npx dockerfile-x` (this command line is documented using `--help`). If new features are added to docker/dockerfiles as new keywords etc... they will be supported without the need to make change on this lib. So you have to precompile you dockerfile, you can think it as a template engine specially designed for Dockerfiles ;-)

**Comment by ulgens** [comment]
> @jessfraz closed this as completed on Jul 10, 2015

This is not completed 🙁

**Comment by rgpublic** [comment]
@ulgens : I think you are out of luck. This won't be solved in the near future. Same tedious discussion as for docker. It seems there's some ideological resistance coming from some of the relevant maintainers. The very absurd argument seems to be: "If we implemented INCLUDE, the Dockerfile would become a full programming language". Or sth. along these lines. So many people have asked for this, voted for this etc. Waited all those years. So many obscure scripts, workaround solutions etc. It's one of those weird things that sometimes can happen to Open Source software. It only needs one reluctant maintainer who doesn't want to implement what everyone asks for and what would only need two lines of code and that's it. You could only fork. But it's not worth it. Like many others I'm now using a simple preprocessor to fix this.

**Comment by ulgens** [comment]
@rgpublic I wasn't asking to fix it, I'm saying that issue's current state is misleading. It should be closed as "won't do" (if that's the last decision).

**Comment by codethief** [comment]
Not really relevant for the present ticket but maybe still of interest to some: I have wanted INCLUDE for the longest time but recently gave up on Dockerfiles entirely for building container images and have since been using [Dagger](https://dagger.io/) to define images through Python code (which, of course, offers a proper import system, type checking, et cetera). [Example code](https://docs.dagger.io/quickstart/ci?sdk=python#construct-a-pipeline).¹

In my experience, for complicated builds (lots of dependencies, multiple architectures, and multiple stages, etc.) Dockerfiles just become unmaintainable at some point, especially if you also want to avoid busting the image layer cache prematurely. Dagger, in turn, has been working pretty well so far, even if it is still somewhat experimental and can be a little rough around the edges. (I'm not associated with Dagger at all, just a fan of their work.)

¹) Don't let yourselves get distracted by Dagger's LLM/agentic features: They can be useful, too, but the core of Dagger is just an SDK/engine around BuildKit to build & run containers programmatically.

**Comment by rmnilin** [comment]
```Dockerfile
# base.Dockerfile
FROM alpine:3.22.1
RUN apk add --no-cache curl
```

```Dockerfile
# app.Dockerfile
FROM base
CMD ["curl", "-s", "https://httpbin.org/base64/SGVsbG8sIHdvcmxkIQo="]
```

```hcl
# docker-bake.hcl
target "base" {
  dockerfile = "base.Dockerfile"
}
target "app" {
  dockerfile = "app.Dockerfile"
  contexts = {
    base = "target:base"
  }
  tags = ["app"]
}
```

```shell
$ docker --version
28.3.3
$ docker bake app
[+] Building... FINISHED
$ docker images --all
REPOSITORY                        TAG                    IMAGE ID       CREATED         SIZE
app                               latest                 f929523c8f69   1 second ago    13.8MB
$ docker run --rm app
Hello, world!
```

I recently encountered this problem in our enterprise monorepo and found these articles linked by @creative-resort above:
- https://docker.com/blog/dockerfiles-now-support-multiple-build-contexts/#:~:text=Create%20Build%20Pipelines%20by%20Linking%20bake%20Targets
- https://docs.docker.com/build/bake/contexts/#using-a-result-of-one-target-as-a-base-image-in-another-target

The articles are great for learning about modern Docker features, but I felt they were a bit bloated when I just wanted to solve the `INCLUDE` problem, so here is my simple solution example.

Compared to this issue's original goal, this solution has significant drawbacks. You now need to be familiar with the HCL configuration language and the bake file schema. This adds considerable complexity compared to a simple `INCLUDE` keyword. Ironically, this introduces an entire configuration language to orchestrate builds, while the original `INCLUDE` proposal was resisted by some for fear it would "turn Dockerfile syntax into a scripting language" and "lead to bad coding practices."

Also it still introduces intermediate images, which can be a pro or con depending on the context. As shown, the base image from `target:base` doesn't clutter `docker images` by default because it's stored only in BuildKit's isolated cache, not the main Docker daemon's image store.

But it has its own huge pros as well, e.g. around caching.

---

Update in response to the comment below:

Special thanks to the engineers who provide syntax extensions and other solutions to this problem. Unfortunately, some projects cannot integrate additional third-party solutions due to philosophy, strict requirements, technical limitations, etc. The baking solution above focuses on what is available out of the box. Of course, a third-party solution can be the clear winner in other cases.

**Comment by devthejo** [comment]
Hello @rmnilin, I don't know if you see my previous message, it's maybe lost in the mass of messages, so I repost it

I developed this buildkit custom syntax frontend plugin: [dockerfile-x](https://github.com/devthefuture-org/dockerfile-x)

just one line at the top of file `# syntax = devthefuture/dockerfile-x` and that's it !

 ```
# syntax = devthefuture/dockerfile-x

FROM ./base/dockerfile

COPY --from=./build/dockerfile#build-stage /app /app

INCLUDE ./other/dockerfile
```
